[Clang] add wraps and no_wraps attributes

JustinStitt opened the pull request:
> [!NOTE]  
> This PR is an expansion of https://github.com/llvm/llvm-project/pull/86618, with a larger scope including `no_wraps` and [SSCL](https://clang.llvm.org/docs/SanitizerSpecialCaseList.html) integration. There's more background and review that is worth reading in #86618 as well.

Implement `__attribute__((wraps))` and `__attribute__((no_wraps))`. These attributes are used to annotate the expected wrap-around behavior for types, variables, struct members, or function parameters.

These attributes serve as convenient in-source annotations for the expected overflow behavior of code. In addition, these attributes also have functional changes when used alongside certain sanitizers or warnings.

The `signed-integer-overflow`, `unsigned-integer-overflow`, `implicit-signed-integer-truncation` and the `implicit-unsigned-integer-truncation` sanitizers have different behavior when interacting with the `wraps` attribute.

This is most useful for users that want to enable overflow or truncation sanitizers but also have some code they want to not be instrumented. Without this PR, they can only achieve translation unit level or at best function level granularity without changing arithmetic expressions directly. In an effort to not make basic arithmetic rely on wrappers like `__builtin_add_overflow()`, this PR introduces type and variable level granularity.

#### Example
```c
void foo(int __attribute__((wraps)) A) {
  ++A; // No overflow/truncation sanitizers will instrument this code
}
```

#### no_wraps

`no_wraps` is interesting because it reflects the default case for integer operations and the associated sanitizers. In the general case, `no_wraps` just serves as a friendly annotation to readers of code that something is really not supposed to wrap-around. The sanitizers will instrument them all the same as they would if this attribute was not present.

What makes `no_wraps` useful is when [SSCLs](https://clang.llvm.org/docs/SanitizerSpecialCaseList.html) are used alongside them. https://github.com/llvm/llvm-project/pull/107332 introduced type-filtering for the overflow and truncation sanitizers which can be overriden by `no_wraps` and `wraps`. To illustrate, let's look at another example:


```
[signed-integer-overflow]
type:*
```

```c
typedef int __attribute__((no_wraps)) non_wrapping_int;

void bar(int A, non_wrapping_int B) {
  ++A; // will not have instrumentation (ignored within the ignorelist)
  ++B; // will have instrumentation (ignored within ignorelist, but overridden by no_wraps)
}
```

----
### CC
@vitalybuka @kees @melver @bwendling

JustinStitt committed:
add wraps, no_wraps attributes

JustinStitt committed:
fix typo in docs

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
fix more language in docs

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
run formatter

Signed-off-by: Justin Stitt <justinstitt@google.com>

vitalybuka commented on clang/docs/SanitizerSpecialCaseList.rst:
```diff
@@ -67,9 +67,11 @@ types specified within an ignorelist.
     int a = 2147483647; // INT_MAX
     ++a;                // Normally, an overflow with -fsanitize=signed-integer-overflow
   }
+
```
Please do unrelated fixes in separate PRs.

JustinStitt replied:
The wraps docs reference the SSCL docs and I fixed some whitespace in the example to make it more readable.

vitalybuka replied:
Yes, I'ts about spaces. Even if it's useful change  just create a new PR. Even from web UI. If you had commit access, no review is needed.

JustinStitt replied:
I don't have commit access, but I'll revert that whitespace change 👍

vitalybuka commented on clang/include/clang/AST/Expr.h:
```diff
@@ -4142,6 +4142,12 @@ class BinaryOperator : public Expr {
+  bool hasWrappingOperand(const ASTContext &Ctx) const;
+
+  /// How about the no_wraps attribute?
+  bool hasNonWrappingOperand(const ASTContext &Ctx) const;
```
This is doc string for https://llvm.org/doxygen/classllvm_1_1BinaryOperator.html, so it's better to be independent.

JustinStitt committed:
use standalone doc string and reinsate removed newline

This commit is a combination of two commits. I'm force pushing because
the build bot had some problems with an (empty?) or whitespace-only
commit during checkout. Hopefully it takes this one.

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
simplify parsing by limiting to TypeAttr

Changing the attr base class from DeclOrTypeAttr to TypeAttr allows us
to better isolate the attribute handling logic just in SemaType.cpp in
the handleWrapsAttr() method. Furthermore, check for language
compatibility at this step as well.

Rely on `COnly` to provide the warnings to the user, then silently
discard the attribute.

|	let LangOpts = [COnly];

Signed-off-by: Justin Stitt <justinstitt@google.com>

melver commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -8710,3 +8710,103 @@ Declares that a function potentially allocates heap memory, and prevents any pot
+If a type or variable is attributed by both ``wraps`` and ``no_wraps``, then
+``no_wraps`` takes precedence -- regardless of the order of attribution.
+
+Note that ``no_wraps`` makes no guarantees about the definedness of arithmetic
```
This sentence is confusing. Is that implied that if I add `no_wraps`, that overflow is now UB? Why does fwrapv help here?

In general, I think there need to be more explicit mentions of -fwrapv and how these attribtues interact with these flags - afaik the attributes give us selective fwrapv and fno-wrapv respectively.

JustinStitt replied:
The "definedness" of signed overflow is handled at the language level with things like `-fwrapv` or `-fno-wrapv`.

Usually codebases opt for `-fwrapv` to avoid eager optimizations of UB arithmetic, think of stuff like:

```c
if (a + 1 < a) { ... }
```

which might be completely optimized away without `-fwrapv`.

I didn't really want things marked with `no_wraps` to go against what users have specified with `-fwrapv` and their code starts getting optimized away. `wraps` and `no_wraps` are intended to be used to modify sanitizer instrumentation and not entire language properties.

If `no_wraps` went against `-fwrapv` then I know for a fact the kernel would not use it, and we really need it over there.

JustinStitt commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -433,6 +433,26 @@ Attribute Changes in Clang
+  when paired with `Sanitizer Special Case Lists (SSCL)
+  <https://clang.llvm.org/docs/SanitizerSpecialCaseList.html>`_.
+
+  These attributes are only valid for C, as there are built-in language
```
No, it won't break the build. You will get an ignored attribute warning and the attributes won't be applied or have any effect. (I'll document this).

These attributes are relatively well supported in C++ but that language has many more ways for the attribute to disappear unexpectedly. 

This attribute is mostly needed in C (as there aren't any alternatives).

JustinStitt replied:
I guess it wouldn't be the worst idea to just let the attribute exist in C++ and document common breakages? We should really urge C++ users to use the type system and operator overloading to their advantage, though, and not use wraps/no_wraps.

What do we think?

melver replied:
My view is that the semantics of C code should, where possible, not change when compiling as C++. This also holds for your type-based proposal.

vitalybuka commented:
I am comfortable accepting sanitization related stuff, and will be happy to look into implementation details.
To me this approach is LGTM, but this patch goes beyond sanitizers into the area of tuning behavior of UB code,
so I think we need feedback from maintainers of the related code @erichkeane, @efriedma-quic.

vitalybuka reviewed (commented):
I guess we already discussed that before. My preference is the wrap and sanitize attributes are interdependent. Open to be convinced otherwise.

But seems like we want consistency with https://godbolt.org/z/crhdaczx1

* `__attribute__((wraps,no_sanitize("signed-integer-overflow")))` wraps quietly
* `__attribute__((wraps))`  wraps, but may report if runs with `-fsanitize=signed-integer-overflow`
* `__attribute__((no_sanitize("signed-integer-overflow")))` reports, but in recovery mode, compile UB any way it likes.


Note: as-is __attribute__((no_sanitize("signed-integer-overflow"))) is not supported on variables, but if we can do wraps, we can do no_sanitize.

JustinStitt committed:
undo change to whitespace

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -433,6 +433,26 @@ Attribute Changes in Clang
+  There is also ``__attribute__((no_wraps))`` which can be added to types or
+  variable declarations. Types or variables with this attribute may be
+  instrumented by overflow sanitizers, if enabled. Note that this matches the
+  default behavior of integer types. So, in most cases, ``no_wraps`` serves
```
When used alongside very strict SSCLs there is great utility to be gained from `no_wraps`.

In the kernel we want to use an SSCL ignorelist like this:
```
[{signed-integer-overflow,unsigned-integer-overflow}]
type:*
```

then annotate specific types like `size_t` in source:

``` cpp
typedef __kernel_size_t    size_t __attribute__((no_wraps));
```

JustinStitt committed:
add similarity to -fwrapv to docs

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -8710,3 +8710,103 @@ Declares that a function potentially allocates heap memory, and prevents any pot
+attributed with the ``wraps`` attribute.
+
+The ``signed-integer-overflow``, ``unsigned-integer-overflow``,
+``implicit-signed-integer-truncation`` and the
```
The entire intention of `wraps` is to _disable_ instrumentation entirely. We never want a report or trap out of such expressions. Developers who have very intentional wrapping arithmetic should be able to delegate types as "wrapping" and move on with their day.

Furthermore, most of the use of `no_wraps` is in conjunction with SSCLs to further control overflow behavior at the type-level. Sanitizer recovery modes remain separate from all of this.

JustinStitt commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -8710,3 +8710,103 @@ Declares that a function potentially allocates heap memory, and prevents any pot
+  [signed-integer-overflow]
+  type:*
+
+``no_wraps`` can override the behavior provided by the ignorelist to
```
Read my other comments about `no_wraps` + SSCLs and recovery modes. I don't think any consideration to recovery modes should be made from `wraps` or `no_wraps`. These attributes only care about sanitizer instrumentation and _not_ the handling thereof.

JustinStitt replied:
But I see what you're saying about `__attribute__((sanitize(...)))`. Do you think we should use `__attribute__((sanitize(...)))` instead of `__attribute__((no_wraps))`? I think this may become confusing to users who think this is turning on a sanitizer, which is not what it does; It really is stating that something is "sanitizeable".

And for `wraps`, maybe you prefer `__attribute__((no_sanitize(...)))` over `__attribute__((wraps))`?

vitalybuka replied:
> But I see what you're saying about `__attribute__((sanitize(...)))`. Do you think we should use `__attribute__((sanitize(...)))` instead of `__attribute__((no_wraps))`? I think this may become confusing to users who think this is turning on a sanitizer, which is not what it does; It really is stating that something is "sanitizeable".
> 
> And for `wraps`, maybe you prefer `__attribute__((no_sanitize(...)))` over `__attribute__((wraps))`?

Yes, if, like you said before, wraps is about disabling sanitizers, than `no_sanitize` looks better. We already can apply no_sanitizer(address) on some globals.

So, no_wraps(or sanitize) is here just to inverse ignore list for particular variable?
I don't think it's worth of it.

In general idea of sanitizers is to check code with minimal modification.
Some stuff required to avoid false-positives, but false-negatives we usually quite easy on them for sake of usability.
So if you need to suppress type, I don't think cherry-picking particular variables to suppress will have wide adoption, and may only annoy opposing users even more. 

Saying from my experience in google3, I would rather over-suppress than introduce additional burden.

BTW. you have "typedef" support in ignore list. Wouldn't e nicer just to change e.g type from e.g. int to sanitizer_int typedef or something?

So I see value in wraps/no_sanitize, but the reverse looks questionable.

vitalybuka replied:
Patch itself, if we introduce wrap/no_sanitize, I don't think no_wrap/sanitize is a big problem for clang maintainers. They may think otherwise :)

However, this just smells like YAGNI to me.

I would see Linux roll-out, as multi step process:
1. Roll-out with over-suppression: ignore list and no_sanitizer attribute.
2. If Linux is able to keep up with those checks, maybe reduce suppression scope with available tools.
3. Then you may want to go further and need no_wrap/sanitize attribute. But I suspect ROI from step 3 will be so low, that it's not going to happen.

JustinStitt replied:
> 1. Roll-out with over-suppression: ignore list and no_sanitizer attribute.

The current plan for the Kernel is ignoring all types and then using `no_wraps` or `sanitize` (the name of the thing doesn't matter). This helps towards (2)

> 2. If Linux is able to keep up with those checks, maybe reduce suppression scope with available tools.

We will be able to keep up with the checks because we will iteratively expand the scope of `no_wraps` to more types in accordance with what our overflow-fixing bandwidth is.

> 3. Then you may want to go further and need no_wrap/sanitize attribute. But I suspect ROI from step 3 will be so low, that it's not going to happen.

Integer overflow (the unexpected kind) cause a lot of bugs and are an entire attack surface of their own. If we can provide more tools to developers such that enabling sanitizers makes more sense and causes less headaches (false-positives) then I see the ROI being potentially high.

This is what `wraps` or `no_wraps` does. It will help codebases better handle wrapping arithmetic and suddenly it becomes less _magic_ and more _safe_

JustinStitt replied:
@vitalybuka  Look at AOSP for example: https://android-review.googlesource.com/c/kernel/common/+/3343205

> Sanitizing signed integer overflow in Linux kernel might be useful to
> spot bugs, however there are cases where overflow is intended, for
> example, `atomic_long_t` type and such cases can't be annotated to not
> trigger UBSAN.

They, and other kernel projects that use sanitizers, really want a wraps or wraps-equivalent _thing_.

vitalybuka replied:
> The current plan for the Kernel is ignoring all types and then using no_wraps or sanitize (the name of the thing doesn't matter) on a few types. This helps towards (2.)

I see, you are doing this in opposite way. My intuition it's more complicated than needed. But I didn't work on Kernel, maybe there are own specifics. It looks like you need to allow all types used in kernel one by one in the end.

We tried allow list like approach in google3. It usually goes slower then ignore list approach - sanitizer everything that pass, the rest go in ignore list, usually by src: or fun:. Then next stage shrinking ignore list by fixing or converting to attribute on functions.

@ramosian-glider  has experience with other sanitizers, maybe you have opinion on the approach https://github.com/llvm/llvm-project/pull/115094#discussion_r1839130783?

JustinStitt commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -8710,3 +8710,103 @@ Declares that a function potentially allocates heap memory, and prevents any pot
+  let Category = DocCatField;
+  let Content = [{
+The ``wraps`` attribute can be used with type or variable declarations to
+denote that arithmetic containing attributed types or variables have defined
```
Yes, good idea.

JustinStitt commented on clang/test/Sema/attr-wraps.c:
```diff
@@ -0,0 +1,48 @@
+// RUN: %clang_cc1 %s -verify -fsyntax-only -triple x86_64-pc-linux-gnu
+typedef int __attribute__((wraps)) wrapping_int;
```
> Can we test more types?

Yes, I'll add some more tests 😄 

> 
> Can the attributes be applied to pointers? Both the pointer itself and the pointee? It's worth adding tests, just to ensure nothing regresses here.

Pointers no:
```
warning: using attribute 'wraps' with non-integer type 'char *' has no function and is potentially misleading [-Wuseless-wraps-attribute]
    2 |   char *A __attribute__((wraps));
```

pointees yes:
```
  char B __attribute__((wraps)) = 127;
  char *A = &B;
  (++B); // no -fsanitize=implicit-signed-integer-truncation splat!
```

JustinStitt committed:
better discard of useless wraps attr

Signed-off-by: Justin Stitt <justinstitt@google.com>

erichkeane commented on clang/lib/AST/Type.cpp:
```diff
@@ -2863,6 +2863,15 @@ bool QualType::isWebAssemblyFuncrefType() const {
 }
 
+bool QualType::hasWrapsAttr() const {
+  return !isNull() && getTypePtr()->hasAttr(attr::Wraps) &&
```
we probably want to do 'more work' to pull out the type you want/canonicalize the type here.

erichkeane commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -10818,11 +10819,39 @@ static bool IsImplicitBoolFloatConversion(Sema &S, Expr *Ex, bool ToBool) {
   for (unsigned i = 0; i < NumArgs; ++i) {
     Expr *CurrA = TheCall->getArg(i);
+
+    if (IsWrapsAttrLost(S, TheCall, FD, i))
```
This is looking an AWFUL lot like this needing to be a type with a defined conversion.  I don't think `AttributedType` is sufficient here, this should also probably be a diagnosed-error, which you get when it is a part of the actual type.

JustinStitt replied:
> This is looking an AWFUL lot like this needing to be a type with a defined conversion. 

Do you mean make `wraps` or `no_wraps` a type in the same way that [ext_vector_type](https://clang.llvm.org/docs/LanguageExtensions.html#vectors-and-extended-vectors) works? Or even BTFTagAttributedType?

erichkeane commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -6536,6 +6536,30 @@ static void HandleBTFTypeTagAttribute(QualType &Type, const ParsedAttr &Attr,
+
+  // wraps and no_wraps are most useful and consistent when used with C. Other
+  // languages have better alternatives within their type systems.
+  if (S.LangOpts.CPlusPlus || S.LangOpts.ObjC)
```
This should diagnose.

JustinStitt replied:
I'm relying on the default diagnostic provided from the LangOpt `COnly` in the table gen. The message is the following:

```
test.cc:2:28: warning: 'wraps' attribute ignored [-Wignored-attributes]
    2 | typedef int __attribute__((wraps)) wrapping_int;
      |                            ^
1 warning generated.
```

erichkeane replied:
Ah!  Right... do we need this check then?  Or can this be an assert?

JustinStitt replied:
For whatever reason, the execution still ends up here during parsing. Can't be an assert :(

erichkeane reviewed (commented):
It seems to me that this should just modify the type, rather than risk getting lost like it does.  If you want to define the conversion to the underlying integer type(likely via cast!) that is sensible as well, but this is going to end up getting lost pretty easily, so likely worth trying to see if this is worht a new type in the AST.

Also, did we have an RFC on this?  I'd like to see what the community at large has to say.

JustinStitt commented on clang/lib/AST/Expr.cpp:
```diff
@@ -4852,6 +4862,10 @@ BinaryOperator::BinaryOperator(const ASTContext &Ctx, Expr *lhs, Expr *rhs,
   if (hasStoredFPFeatures())
     setStoredFPFeatures(FPFeatures);
   setDependence(computeDependence(this));
+  if (hasWrappingOperand(Ctx))
```
Implementation-wise, it is OK for both attributes to exist on the same type.

`hasWrapsAttr()` cautiously checks for `no_wraps` before claiming we have `wraps`.
```c
bool QualType::hasWrapsAttr() const {
  return !isNull() && getTypePtr()->hasAttr(attr::Wraps) &&
         !getTypePtr()->hasAttr(attr::NoWraps);
}

```

But yeah you're right, this is confusing because if you look at an ast dump you'll see both attributes applied to the type and it may not be obvious which has precedence. I'll add a diagnostic and a sane default of keeping the `no_wraps` and dropping `wraps` for cases like this.

JustinStitt replied:
As of https://github.com/llvm/llvm-project/pull/115094/commits/445f8c7c93484a5476d85e05a289e609e3f38a94, only one attribute will persist through to BinaryOperators. Thanks

JustinStitt commented:
@erichkeane 
> Also, did we have an RFC on this? I'd like to see what the community at large has to say.

Yes, here's the [RFC](https://discourse.llvm.org/t/rfc-add-wraps-attribute-for-granular-integer-overflow-handling/77670) and the [previous PR](https://github.com/llvm/llvm-project/pull/86618) (of which the scope has been expanded a bit so it deserves its own new PR, IMO)

JustinStitt commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6664,6 +6664,13 @@ def err_builtin_counted_by_ref_invalid_lhs_use : Error<
 
+def warn_wraps_attr_var_decl_type_not_integer : Warning<
+  "using attribute '%select{wraps|no_wraps}0' with non-integer type '%1' has no function and is potentially misleading">,
+  InGroup<UselessWrapsAttr>;
```
how about dropping "has no function and" so the final sentence becomes something like:

`"using attribute 'wraps' with non-integer type '%1' is potentially misleading"`

erichkeane replied:
something more like:
`attribute 'wraps' has no effect with non-integer type 'float'; attribute ignored` ?

JustinStitt replied:
since this is now an error ([3de3561](https://github.com/llvm/llvm-project/pull/115094/commits/3de3561f6c6e4e73ae1f5cda3a27a6aff5bf9904)) we can be a bit more assertive with the language. I went with:

`cannot use attribute 'wraps' with non-integer type 'float'` which is an error when parsing.

JustinStitt committed:
-Wuseless-wraps-attribute is now an error

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
add error for using both wraps and no_wraps together

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
only persist one of wraps or no_wraps

... and prefer no_wraps if both want to exist on the same binary expression

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -8710,3 +8710,103 @@ Declares that a function potentially allocates heap memory, and prevents any pot
+be automatically applied to the result type of arithmetic expressions
+containing a wrapping operand.
+
+If a type or variable is attributed by both ``wraps`` and ``no_wraps``, then
```
Since [093bb7f](https://github.com/llvm/llvm-project/pull/115094/commits/093bb7fe7a4bb86bb66c79d1173bc06e7c67c5e6), there is now an error when attributing both `wraps` and `no_wraps` to something:

```c
test.c:7:45: error: attribute 'wraps' cannot be used alongside 'no_wraps'
    7 |   int __attribute__((wraps)) __attribute__((no_wraps)) A;
```

... and also there is no more "coming together naturally" of these attributes either. BinaryOperators will prefer capturing `no_wraps` over `wraps` and can only have one or the other. So, disregard my comment above.

JustinStitt committed:
add more truncation tests

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6664,6 +6664,13 @@ def err_builtin_counted_by_ref_invalid_lhs_use : Error<
 def err_builtin_counted_by_ref_has_side_effects : Error<
   "'__builtin_counted_by_ref' argument cannot have side-effects">;
 
+def warn_wraps_attr_var_decl_type_not_integer : Warning<
```
Done in [3de3561](https://github.com/llvm/llvm-project/pull/115094/commits/3de3561f6c6e4e73ae1f5cda3a27a6aff5bf9904). Thanks

JustinStitt committed:
small updates to docs

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
more doc fixups

Signed-off-by: Justin Stitt <justinstitt@google.com>

vitalybuka commented:
Naming wise I prefer `wrap` -> `no_sanitize("undefined")` already exists but does not apply to variables and types

`no_wrap` -> `sanitize("undefined")` does not exist, and a little confusing (no_wrap as well). It may look it enabled sanitization.

Would it make sense to support attributes like  `__attribute__((no_sanitize("-undefined")))` ? Seems like better representation of the fact that we want to undo sanitizer suppression. @erichkeane  WDYT

vitalybuka commented:
> @vitalybuka
> 
> > Would it make sense to support attributes like `__attribute__((no_sanitize("-undefined")))` ? Seems like better representation of the fact that we want to undo sanitizer suppression.
> 
> Is there some other precedence for this `-undefined` syntax with attributes?

No.

> 
> I am working on the PR for supporting `__attribute__((no_sanitize()))` for types. Should the opposite of `no_sanitize` be `sanitizable`? I agree that `sanitize` is no good as it suggests it will enable sanitization, not permit it.

sanitizable is LGTM

> 
> Once I have that PR up can compare it to this PR to find the best way forward.

JustinStitt commented:
RFC regarding canonical wrapping/non-wrapping types in Clang: https://discourse.llvm.org/t/rfc-clang-canonical-wrapping-and-non-wrapping-types/84356

Ultimately, a type like what the RFC describes would supersede this PR in terms of feature completeness and usefulness. I'll close this PR if that RFC suggests a new direction is required.

pings:
@erichkeane pinging you because you asked about an RFC.

melver commented:
> RFC regarding canonical wrapping/non-wrapping types in Clang: https://discourse.llvm.org/t/rfc-clang-canonical-wrapping-and-non-wrapping-types/84356
> 
> Ultimately, a type like what the RFC describes would supersede this PR in terms of feature completeness and usefulness. I'll close this PR if that RFC suggests a new direction is required.
> 
> pings: @erichkeane pinging you because you asked about an RFC.

+1 - I'd be in favor of the type-based approach. It does make more sense that this is a type qualifier, vs. an attribute which the compiler might loose as soon as type conversions or take-pointer operations are involved.

JustinStitt commented:
superseded by https://github.com/llvm/llvm-project/pull/148914.

closing this.
